### ISSUE TITLE
App sometimes stops loading tiles from local mapnik tile cache

### ISSUE THREAD

**Issue by isync** [geopaparazzi/geopaparazzi#289]
(I'm not sure if this is a geopaparazzi defect, or in one of the underlying libraries.)

When connected to the Internet, I prefetch tiles for unconnected/on-the-go usage by zooming into areas I'm going to. They properly end up in the local sqlite cache file and when I disconnect my device, the prefetched tiles appear, but sometimes (when I leave map view, or when I had my device locked, or switched apps, or whatever - when I re-enter map view after this) it seems the app somehow "forgot all tiles" and I'm presented with an empty/grey canvas.
Restarting the whole app remedies that, but that's not really an option while logging GPS..., right?

(Please excuse my perspective of an inexperienced geopaparazzi user, and my total ignorance of your codebase)
__Would it be possible to add something like a "refresh tiles" button?__

**Comment by mj10777** [comment]
I believe using the `cache` logic is based on the Android system settings
* which is not set or influenced by geopaparazzi

---
Our main wiki page is at:
https://github.com/geopaparazzi/geopaparazzi/wiki

This chapter deals the matter that interests you
https://github.com/geopaparazzi/geopaparazzi/wiki#custom-tiles-based-maps-local-and-remote

---
At the moment I will assume you are using one of the *default*
* `.mapurl'  files

When `geopaparazzi` starts
* it will look through the *maps*  directory
   * and collect all information defined Raster/Vector sources

So you can make a copy of any existing `.mapurl`
* and place it anywhere **inside** the *maps*  directory
   * it will then be read when `geopaparazzi` starts

This `.mapurl` can then be adapted to your needs
* this is a simple text file
   * which you adapt with an external editor
   * there is no way to do this from inside `geopaparazzi`

The syntax used for the `.mapurl` can be found here:
https://github.com/geopaparazzi/geopaparazzi/wiki/.mapurl-parameters
(which should be a complete list with samples)

The chapter `Adding mbtiles to Tile sources`, is what you are looking for:
https://github.com/geopaparazzi/geopaparazzi/wiki#id53

You can the define
* where the `.mbtiles` file should be stored inside the *maps*  directory
   * can be different from the `.mapurl`
* the source from where the tiles should be downloaded from
* the area [bounds] that should be downloaded
* the zoom levels to be downloaded

>> I prefetch tiles for unconnected/on-the-go usage by zooming into areas I'm going to

replacing what you are doing.

When `geoparazzi` starts
* it will read this file
   * will create the `.mbtiles` file
   * will create a `request_url` table
      * and fill it with commands needed to fill the `.mbtiles` files with the requested tiles
   * will then start to download each entry in the `request_url` table
      * removing each record **after** the tile has been stored in the `.mbtiles` file
      * it will also check if the tile already exists before downloading

This is done as a **background** task
* should `geoparazzi` be stopped for any reason
   * it will **resume** the task when restarted

---
This `.mapurl` (with a `.mbtiles` file defined) can be opened at anytime
* you can 'wander' around
   * for any tile being shown
     * it will first check if the tile already exists
        * if not, it will be downloaded and stored in the `.mbtiles`
     * it will then display the tile

So you have 2 ways in filling the `.mbtiles` file
* using the `request_url` logic
   * will insure that the area is complete

--
When the `request_url` table is empty
* it will be deleted

Filling `request_url` is the first task
* so after that has been completed
   * the only way to know this is to look at the log entry of android
   * or the 'log' table inside `geopaparazzi.gpap` (which is a sqlite Database)
     * the will be messages that the `request_url` has been 
        * filled
        * in 10% steps show many tiles have been downloaded / are open
        * the download task has been completed
* the `.mapurl` can be changed
   * **adding** new entries into the `request_url`

Please note that Android only supports file
* up to 2.147.483.648 bytes (2 Gigabyte)

---
The `.mbtiles` file can also be opened directly
* but then missing tiles will **not** be downloaded

So once everything has been downloaded
* you can delete the `.mapurl` file

**Comment by moovida** [comment]
Hi @mj10777 , I experienced exactly the problem that was reported. I also reported this back some time ago but thought it was fixed.
To reproduce this it should be just be as simple as: 

* use one of the defaut mapruls and downoad some tiles
* **turn off any internet connection**. This is the point. If internet is on, it will load the tiles quickly, hence using the cache, but if internet is off, it won't load the tiles
* try to load the same region offline

**Comment by mj10777** [comment]
>> default mapurls 

without mbtiles ? (I assume)

>> cache
* is done from Android?

**Comment by moovida** [comment]
The default mapurls all should cache. Simply use mapnik. 

We do not use any android caching engine.

**Comment by isync** [comment]
It seems 2 things need clarification:

1. I'm not talking about Android cache, but the "cache" file Geopaparazzi obviously creates. It's located in _maps/defaulttiles/_mapnik.mbtiles_. This file is growing when I use the shipped mapnik.mapurl file. It works, it stores titles I already looked at in mapview and obviously works as a mapsource when I go offline. Except for _sometimes_, when I experience @moovida 's observation that this sometimes fails (although I can't reliably reproduce it, just try to make sense of what I described above (after exiting mapview, after locked-screen, etc.))

2. Thank you @mj10777 for your detailed description. As it seems, my default/shipped mapurls file already has this configured out-of-the-box:
> url=http://tile.openstreetmap.org/ZZZ/XXX/YYY.png
> minzoom=0
> maxzoom=19
> center=11.42 46.8
> type=google
> format=png
> defaultzoom=13
> mbtiles=defaulttiles/_mapnik.mbtiles
> description=Mapnik - Openstreetmap Slippy Map Tileserver - Data, imagery and map information provided by MapQuest, OpenStreetMap and contributors, ODbL.

I wasn't aware of the ```request_url``` and ```request_type``` (run,load,fill) parameters. This can in fact replace what I am doing manually by "wandering around" - although I have to see how this works, working with center= and bounds= (rectangular areas vs. polygonal), but I'll try that

Take aways:

* Part of this here stems from Geopaparazzi's documentation being split; it's on the wiki, while the main README.md is pointing users to the outdated docs at http://geopaparazzi.github.io/geopaparazzi/
* Idea: What about an UI allowing the user to draw a polygon on the map to define an area for the request_url logic to pre-populate the mbtiles file with?
* __All this does not answer why mapview sometimes seems to "forget" about tiles being stored in the mbtiles files configured in, for example, mapnik.mapurl__ (see my above paragraph 1.)

**Comment by tghoward** [comment]
I have not had good success with this caching, so it must be inconsistent
somehow. This would be *great* if it would work consistently.   But there
should also be a way to clear the cache for specific layers if disk space
is getting too full. Perhaps there already is?

**Comment by mj10777** [comment]
>> I have not had good success with this caching

Do you mean using `.mapurl` together with `.mbtiles` ?
* I am getting the impression that it is this **combination** that is causing the problem 

>> to clear the cache for specific layers

If the above **combination** is meant by this
* deleting the `.mbtiles` file defined in the `.mapurl` would deal with this

>> Perhaps there already is?

We have always **avoided** the direct manipulation of non-geopaparazzi data
* the assumption being, that this data **belongs** to the user
   * the user themselves should make this decision
      * so called `intelligent` solutions
         * mostly leads to **disaster**, in my experience

**Comment by mj10777** [comment]
To @isync :

**Documentation**
* to my knowledge an new documentation has been prepared
   * last touches being done during the summer in preparation for publication
      * but I don't recall what has become of this  
* the `README.md` should definitely point to
   * the main wiki-page
      * and that point to the latest documentation

* the `README.md` 
   * can only be updated through a commit 
   * where as the wiki-page can be updated online 

---

**UI for mapurl request_url logic**
* insofar as for determining the bounds to be downloaded
   * this is a good idea, since it can help avoiding such situations as:
      * downloading all the tiles of the world
         * if `bounds` are **not** set, the whole world is set as default
      * at Zoom Level 18
         * if Max-Zoom  is **not** set, this is what is used
      * map-servers become **very** unhappy
         * when such things happen
   * the above is one of two major reasons for not implementing an **automatic** download
      * the other being that some map-servers **do not permit** 
         * the permanent downloading and storing of there tiles
      * in the past, some applications have been barred from Google-Play
        * for offering just what you requested
   * since there is no way for us to **determine** online
      * i.e. before starting such a download process
      * whether this is permitted or not
         * offering such **automatic** downloads have **not** been offered
   * since there are also map-servers that, within reason, do not object to an **automatic** download
      * the possibility to use the 'mapurl request_url'  logic, where justified, is offered
         * allowing the user to make the decision on when it is needed
   * to my knowledge, there has never be a complaint from any map-servers
     * that their services are being misused

---

**center=** parameter in `.mapurl`
There is a misunderstanding here about the use of the **center** parameter
* this has **nothing** to do with the **request_url**  logic

The idea here is:
* when the user loads a new map
   * and was previously **outside** the range (bound) of that map
      * the start position that is shown when the new map is loaded
        * is the given, user defined, **center** point
           * when **not** set: the center of the map is used

Sample:
* you are using a map of England, wandering around London
   * you then load a map of Italy
      * the MapView would then zoom to the given **center** point
* In the case of Italy
   * it would be wise to set the **center** point to another position (such as Rome)
      * since the center of the Map would be somewhere in the Tyrrhenian Sea
         * at Zoom-Level 17, not very useful
           * unless you are a fisherman
              * using a nautical map for the area between Sardinia and Napels
* the idea is to avoid the user
   * getting lost in a white sea of tiles

**Comment by mj10777** [comment]
Lets try the analyse the obtained information about this problem
* with the goal of resolving the problem
   * for which the **cause** must be determined

---
Question 1:
* Does this **only** happen when `.mapurl` and `.mbtiles` are used together
   * my impression is: **yes**

Question 2: (assuming **yes** to the above)
* based on the following information
   * When a `.mbtiles` file is used **directly** 
      * `MbtilesDatabaseHandler` is used
         * I use this often and such an error has never occurred
   * When a `.mbtiles` file is used from a `.mapurl`
      * `CustomTileDatabaseHandler` is used
         * which uses `CustomTileDownloader`
            * which, when `.mbtiles` file is used
               * uses `MbtilesDatabaseHandler`
                  * **mbtilesDatabase != null**
   * `The default mapurls all should cache. Simply use mapnik.`
      * I assume this to mean that mapnik has its own cache system
         * which **may** be the cause of this error
* When using the mapnik cache system
   * is `CustomTileDownloader.executeJob`
      * being called **before** or **after** the mapnik cache system logic runs?

---
* not yet finished

**Comment by mj10777** [comment]
Based on the instruction of @moovida 
* I used the default mapnick mapurl for the first time yesterday
   * filling in a small area

---
I have just restarted this mapurl
* and the tiles (as expected) showed, filling while moving around

I then turned off the WLAN-Connection
* while the application was still active
   * the existing tiles show
      * with new areas showing blank (as expected)
* after restarting application
   * same result

So, up to now, I **cannot reproduce** this error
* which, however, may only occur when a certain limit has been exceded

**Comment by mj10777** [comment]
After writing 'Question 5b:'. I attempted the following
* with WLAN-off
   * moved to an area where tile were missing
* then turned the WLAN on
   * the tiles were **not** loaded
* moved in a westerly direction
   * that had **not** yet been requested
      * these tiles were filled
         * the tiles, that were requested while the WLAN was off
            * were not

So it seems that the mapsforge-library is **remembering**
* that there are requests that could not be fulfilled

Looking at `CustomTileDownloader.executeJob`
* if an mbtiles-Tile exists
   * it is returned without further checking

So my assumption (at the moment) is that
* under the above condition 
   * 'executeJob' is **not** being called at all
      * thus portions of Questions 5a+b can be answered with **yes**

Since this can be **reproduce**
* I will build in some debuging code to confirm this

**Comment by isync** [comment]
First Mark (@mj10777), I'd like to thank you for putting such tremendous effort into this!

Without knowing about the inner workings you obviously know in and out, for me it seems, from how you describe your reproduction efforts, you are on the right track with a number of your assumptions, in Question 2, Question 5b, (Q 9).
From what I understand, let me paraphrase one of your "possible problem sources": It seems, the mapview can be maneuvered into a state where it thinks it can download tiles from the Internet, but this fails when the device is offline, and when this fails, the Downloader engine stores this failed download and stops the logic from looking further, into the MBTiles file, which in turn would have had this tile, but is never asked for it. Restarting the app clears this "remember failed tile downloads" mechanism, and when Internet is off from the app's startup, the logic behaves a bit differently, and doesn't store "failed net requests" anymore, not preventing the logic from asking the MBTiles file for tiles...  something like that. - - No need to correct me here, if I got something wrong, my verbose paraphrase is just meant to encourage you that I think you are on the right track.

Also, your Question 2 sparked the idea for an __acceptable workaround__, at least for me: simply using the mbtiles cache file directly as a source. As it seems layering mapurl and mbtiles backends is the root of the problem, so I can simply avoid that. I would enable mapurl+mbtiles while connected to the Internet, wander around and build my custom mbtiles file - and when I disconnect, I switch mapsource to the generated mbtiles file and use it directly.

**Comment by mj10777** [comment]
The original default mbtiles file cannot be used
* due to the name ('_' prevents this from showing up)

You can, however, copy this file
* using a name without a leading '_'

Also the mapurl
* with the new file name

This is what I do
* this is why the default mapnik-mapurl has never been used

**Comment by moovida** [comment]
@mj10777 , I am a bit puzzeled that you can't reproduce the problem, since it is reproducable on my device.
I am quite sure this happens in the mapurl code where it should decide to pick the tile from the backed mbtile because it doesn't have any internet connection. I haven't looked in that code for a while, but I remeber a pile of nested ifs. I assume that in that logic something jumps to the wrong conclusion.

@isync we usually place an underscore before the name of the backing mbtiles db, so that it is not seen as an additional source (it would be doubled in case the mapurl's name is the same, which makes sense).  Since this is a bug (and I don't think a difficult one), it should be solved. Anyways, if you want to see the mbtiles source, just name it without underscore as first letter.

**Comment by isync** [comment]
I know about the underscore problem, would use a non-underscore name then in the mapurl file, just didn't mention it for simplicity.
(One seconds guys, I'm in the process of writing up a reproducible path...)

**Comment by moovida** [comment]
> Also the mapurl
> 
> with the new file name
> This is what I do
> 
> this is why the default mapnik-mapurl has never been used

This is **wrong**. The default mapurl doesn't show up, but it is always used. I don't want to see them but want them to be used. The but in the mapurls code should be solved for that.

**Comment by mj10777** [comment]
I meant the created .mbtile file
* that does not show up, at least when I looked this morning

CustomTileDownloader.executeJob
* first checks if the tile exist in the mbtiles file
   and returns if it does (with no other checking)

After that (not in the mbtiles file) it checks if is another file-based file
* and uses that when it is

Then it checks for an internet connection
* and attemts to download it if a connection is active

**Comment by isync** [comment]
I gave it another shot: I can reliably reproduce the behaviour by doing:
0. Have mapurl configured as tile source, with an mbtiles file as fallback/cache
1. Have WLAN on
2. Open the App
3. Open mapview
4. Zoom into an area and wait for tiles to fill
5. Turn off WLAN
6. Exit mapview (go to the Apps overview dashboard)
7. Re-enter mapview, same zoomlevel is remembered, tiles are missing, all is grey
8. Exit the app, close it in task manager
9. Restart the app, go to mapview, tiles are back, at least in parts

**Comment by tghoward** [comment]
I can reproduce the problem with those steps. In step 5, I put my phone
(Nexus 5) in airplane mode.

**Comment by mj10777** [comment]
I left out Step 6
* so I will try that tommorow

**Comment by tghoward** [comment]
And I don't seem to get all the tiles that should have been cached the
first time. By that, I mean, as @isync notes in step 9, the tiles are back,
but only parts of the ones that were downloaded and drawn in step 4.

That's what I was referring to earlier as what seemed inconsistent to me.

**Comment by mj10777** [comment]
You mean that:
* tiles that were requested, but not downloaded, are not there afterwords?

I don't think that this is possible with the normal mapsforge-logic
* the non-fulfilled request has been forgotten

**Comment by tghoward** [comment]
No, I mean that some of the tiles that appear on screen in the map view in
step 4 (so they must have been requested and downloaded), do not appear at
step 9.

**Comment by mj10777** [comment]
Yes, in Step 4 the WLAN is on, therefore has a connection and the results should be stored
* and these do  **not** show up in Step 7, if I have understood this correctly
   * this would mean destroyed (Step  5)
     * and recreated in Step 6
        * the **same** cache is still active

I will try this out tomorrow and see 
* for which tiles CustomTileDownloader.executeJob is called for
   * if any

**Comment by isync** [comment]
@tghoward: I've observed the same and originally wrote in step 9 "tiles are back, more or less" but then edited it to what it says now. As I thought it was my fault, sometimes it's hard to see if a tile was really replaced with the next zoom step. And even after that, it seems it takes some time between state (I make them up here) fetched, displayed, stored, cached, flagged as stored - something like that. And if you exit mapview while some tiles are not "really there" it might result in what you, what I have observed.

It's only a side aspect of the larger issue, I think

**Comment by mj10777** [comment]
Well, this combination now displays the problem correctly
* I have created a log showing what
   * `CustomTileDownloader.executeJob` is doing
      * `[1 -> returning existing mbtiles]`
      * `[2 -> requesting Tile]`
      * `[3 -> saving new mbtiles]`
      * `[4 -> sending blank image]`

For this I will open an new Issue called: 
* `Summery: error: loading tiles from local mapnik tile cache`
   * describing each step
      * showing the (shorten) log portions for each step

**Comment by mj10777** [comment]
The cause of this problem is know known
* the Database connection was lost
   * when leaving the MapView and returning
   * cause **unknown**

**Comment by mj10777** [comment]
A partial solution has been committed to the `rasterlite2_styles` branch
* the mbtiles Database connection will be reestablished when needed

Within the Mapsview
* tile request made **without** internet connection
   * are *sometimes*, but not **always** refreshed
   * new requests will be downloaded (which is strange)

When leaving and return into the MapView
* the missing tiles will be retrieved

This problem will be continued
* in issue #290

and this one closed.

**Comment by isync** [comment]
To everyone involved: a big thank you, for an incredible short turn-around on this issue!

To @mj10777 : you've got a typo in #290 's title: it's "Summary", right?

**Comment by mj10777** [comment]
Yes, was, no doubt, one of many
* corrected
